ERCS/erc-7679.md

```diff
@@ -0,0 +1,223 @@
+     * the executions.
+     * @param smartAccount is the address of the UserOp sender.
+     * @param executions are (destination, value, callData) tuples that
+     * the UserOp wants to execute.  It's an array so the UserOp can
+     * batch executions.
+     * @param context is the data required for the UserOp builder to
+     * properly compute the requested field for the UserOp. 
+     */
+    function getCallData(
+        address smartAccount,
+        Execution[] calldata executions,
+        bytes calldata context
+    ) external view returns (bytes memory);
+    
+    /**
+     * @dev Returns the dummy signature for the UserOp, given the context
+     * and the executions.
+     * @param smartAccount is the address of the UserOp sender.
+     * @param executions are (destination, value, callData) tuples that
+     * the UserOp wants to execute.  It's an array so the UserOp can
+     * batch executions.
+     * @param context is the data required for the UserOp builder to
+     * properly compute the requested field for the UserOp.
+     */
+    function getDummySignature(
+        address smartAccount,
+        Execution[] calldata executions,
+        bytes calldata context
+    ) external view returns (bytes memory signature);
+    
+    /**
+     * @dev Returns a correctly encoded signature, given a UserOp that
+     * has been correctly filled out except for the signature field.
+     * @param smartAccount is the address of the UserOp sender.
+     * @param userOperation is the UserOp.  Every field of the UserOp should
+     * be valid except for the signature field.  The "PackedUserOperation"
+     * struct is as defined in ERC-4337.
+     * @param context is the data required for the UserOp builder to
+     * properly compute the requested field for the UserOp.
+     */
+    function getSignature(
+        address smartAccount,
+        PackedUserOperation calldata userOperation,
+        bytes calldata context
+    ) external view returns (bytes memory signature);
+}
+```
+
+### Using the UserOp builder
+
+To build a UserOp using the UserOp builder, the building party should proceed as follows:
+
+1. Obtain the address of `UserOpBuilder` and a `context` from the account owner.  The `context` is an opaque bytes array from the perspective of the building party.  The smart account implementation may need the `context` in order to properly figure out the UserOp fields.  See the “Rationale” section for more info.
+2. Execute a multicall (batched `eth_call`s) of `getNonce`, `getCallData`, `getDummySignature` with the `context` and executions.  The building party will now have obtained the nonce, calldata, and dummy signature (see “Rationale” for what a dummy signature is).
+3. Fill out a UserOp with the data obtained previously.  This UserOp must be valid except for the `signature` field.  Then call (via `eth_call`) `getSignature` with the UserOp and `context` to obtain a completely valid UserOp.
```

Comment:
How is this envisioned to work exactly? signature are typically generated off-chain by some signer (EOA or 3rd party KMS)

Reply:
`getSignature` is not meant for generating a signature on chain but rather formatting the signature in a way that will be accepted by the account. So signature by the Dapp's session key will still be generated off-chain and can be put in `userOperation.signature` along with the permissionsContext that was passed by the wallet. This on-chain function will just return a fully formatted signature which when entryPoint will pass when calling `validateUserOp`, the on-chain wallet will be able to parse and verify.

Reply:
makes sense!

It seems like smart account contracts will integrate this interface, but if we're querying the account itself, it seems redundant to include `address smartAccount` in all the calls. could you explain whats the use case for that?

Reply:
i didn't get the use case for this function from the naming or comments. i suggest renaming the function to formatSignature, or updating comments to clarify that it takes in an unformatted signature and outputs a formatted one

Reply:
> it seems redundant to include address smartAccount in all the calls. could you explain whats the use case for that?

This interface won't be implemented by the Smart Account itself, but deployed as a standalone contract per Smart Account version. That's why SA address is included as an arg

Reply:
Gotcha, thanks for clarifying!

Reply:
Hmm can you share more about having it in the smart account vs external contract? Accounts are proxies so there shouldn't be any additional init costs. Also, I imagine that looking for this additional contract adds a bunch of logistical overhead whereas if we do the account, wallets/clients/SDKs should already have the address of the account handy